### PR TITLE
Fix issue 22873 - Wrong std.format output for `inout`

### PR DESCRIPTION
**Pull request by pbackus**

### CODE REVIEW

File: std/format/internal/write.d [hunk 1]
```diff
@@ -2071,81 +2069,81 @@ template hasToString(T, Char)
 
     static struct A
     {
-        void toString(Writer)(ref Writer w)
+        void toString(Writer)(ref Writer w) const
```
Comment (JohanEngelen) [inline std/format/internal/write.d:2072]: Are all these new `const` needed? Worries me that this change again changes the Phobos API for the user... (like https://github.com/dlang/phobos/pull/8249 did)

Comment (pbackus) [reply std/format/internal/write.d:2072]: This is not a user-facing API change. These `toString` methods are used only in unit tests. Making them `const` allows them to be called on `inout` struct instances, and makes it possible to test `hasToString!(inout(A), char)` without introducing excessive code duplication (i.e., copy-pasting the entire `unittest` block).

**Comment by schveiguy** [comment]
Interesting error. This is the class that fails `hasToString!(NeuralLayer, char)`: https://github.com/Netflix/vectorflow/blob/master/src/vectorflow/neurallayer.d#L40

Not sure why.

**Comment by pbackus** [comment]
Ugh, I know why it fails. It's this line:

```d
    void init(double random_scale){}
```

...which causes the speculative compilation in this line to fail:

```d
    else static if (is(typeof((T val) { return val.toString(); }(T.init)) S) && isSomeString!S)
```

We assume here that `T.init` is an expression of type `T`, but for this class, its type is actually `void`.

File: std/format/internal/write.d (hunk 1 above)
Comment (JohanEngelen) [reply std/format/internal/write.d:2072]: Of course these `const`s are not user facing. "this change" = this PR.
What I mean is, that the addition of `const` may have been necessary to deal with something that _does_ trickle down to users. (The changes in #8249 also appear not user facing for example, yet they are). It removes some of the tested type space.

Comment (pbackus) [reply std/format/internal/write.d:2179]: Thanks for clarifying. I've restored the test coverage of unqualified `toString` methods.

**Review by schveiguy — APPROVED** [review]

**Comment by schveiguy** [comment]
LGTM, not sure if this shouldn't target stable, since it's a bug fix.

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**